### ISSUE TITLE
Nuget package 6.0.0-beta1 missing LLVM C API symbols on Linux

### ISSUE THREAD

**Issue by tritao** [Microsoft/LLVMSharp#88]
As the title says, the latest Nuget package for `libllvm` is missing C API symbols.

```
joao@joao-HP:~/dev/dotnet/corert/packages/libllvm$ llvm-nm-6.0 5.0.1/runtimes/linux-x64/native/libLLVM.so | grep LLVMModuleCreateWithName
0000000000838d20 T LLVMModuleCreateWithName
0000000000838db0 T LLVMModuleCreateWithNameInContext
joao@joao-HP:~/dev/dotnet/corert/packages/libllvm$ llvm-nm-6.0 6.0.0-beta1/runtimes/linux-x64/native/libLLVM.so | grep LLVMModuleCreateWithName
joao@joao-HP:~/dev/dotnet/corert/packages/libllvm$ 
```

This breaks CoreRT under Linux.

**Comment by mjsabby** [comment]
Thanks @tritao. We'll get this squared away this week, I also noticed a similar break in CoreRT. FYI @morganbr

**Comment by KodrAus** [comment]
@mjsabby is there anything we can do to help fix this up?

**Comment by KodrAus** [comment]
I'm not sure how helpful it would be to anyone, but I'm working around this in `corert` right now by building my own `libLLVM` and patching it in (since `corert` unpacks packages and links the native objects directly).

# On OSX

Clone the `llvm` project:

```shell
$ git clone -b release_60 https://github.com/llvm-mirror/llvm
```

Configure LLVM to build as a single static binary:

```shell
$ mkdir llvm-build
$ cd llvm-build
$ cmake -DLLVM_OPTIMIZED_TABLEGEN=ON -DLLVM_LINK_LLVM_DYLIB=ON -DBUILD_SHARED_LIBS=OFF ../llvm
```

Build LLVM:

 ```
$ make
```

Go for a long walk.

If everything goes well, you should find `llvm-build/lib/libLLVM.dylib`.

**Comment by mjsabby** [comment]
@KodrAus thanks. Working on this.

**Comment by KodrAus** [comment]
Thanks @mjsabby!

I imagine it wouldn't need to change all that much, but do you think it would be worth setting up a repository for `libLLVM` that just contains an Azure Pipeline or something for fetching/building an LLVM binary on the supported platforms, bundling those up into a `nupkg` and publishing to nuget.org? That might give us some more visibility into how the package is produced?

**Comment by mjsabby** [comment]
@KodrAus Yup I so desperately want to do that :)

Happy to create a Microsoft sponsored account if you wanted to take a stab at it.

**Comment by KodrAus** [comment]
Sure :+1: I'm keen to try spike something out. I'll just spin up a repo and see where it goes. I'll open an issue or send a ping if I get anywhere 😃

**Comment by KodrAus** [comment]
So I've actually been able to make some progress on this. Building for OSX and Linux seems to work out ok, but I'm having some trouble getting a single .dll out of the Windows build.

@mjsabby I'm guessing that's what the vcxproj in this repo is all about? Are you able to share some details about how you build LLVM for Windows at the moment?

**Comment by mjsabby** [comment]
Yes the vcxproj is for that. In fact, you need to have the code also be built on Linux and macOS. Because that's what the PInvokes call into since neither platform exposes all the arches.

Are you using a hosted Azure Pipeline instance? Let me know the details and we can put all the relevant info in the repo.

Really appreciate you taking this!

**Comment by KodrAus** [comment]
Ah right. Sure! The repo I'm working on is [here](https://github.com/KodrAus/libllvm), with the current Azure Pipeline looking [something like this](https://dev.azure.com/kodraus/libllvm/_build/results?buildId=53). It's still proof-of-concept quality stuff, but does manage to get the native bits produced in some form on Windows, OSX and Linux.

**Comment by KodrAus** [comment]
I wasn't able to successfully link up a Windows build using the vcxproj from this repo on my machine, but LLVM takes a _really_ long time to build locally on my poor old Windows box (~3hrs vs ~1hr on the build agent), so I might just try running it in the Windows pipeline after LLVM's build and see how I go getting it to link up there.

**Comment by mjsabby** [comment]
@KodrAus do you mind if we copy over your azure pipeline yml over to an MSFT Azure Pipelines? I'll then get beefier VMs for this task.

**Comment by KodrAus** [comment]
No problem! Some beefier agents definitely wouldn't hurt. I think the Windows build that's there now is running close to the disk limit.

I've punted on a couple of design questions in the pipeline though, like whether to pull in the LLVM source as a submodule in the repo or just clone it in CI, just to try prove it was even possible to do so once it's all working we might want to revisit them.

**Comment by KodrAus** [comment]
Success! The Azure Pipeline [I've been working on](https://github.com/KodrAus/libllvm) can now link up a single `.dll` on Windows. So that's Windows, OSX and Linux now built and bundled in a single pipeline. Whether that `.nupkg` I'm assembling actually works is a different question, but all the pieces are there now.

**Comment by KodrAus** [comment]
@mjsabby Just thinking about next steps here, I'm happy to keep chipping away on my `libllvm` repo until it gets to a point where we're sure it works in LLVMSharp.

After that, did we want to transfer the `libllvm` repository somewhere, (is this Microsoft org an appropriate place?), migrate the pipeline to an MS account and set up whatever credentials we'd need to be able to push pre-release packages to MyGet out of CI?

**Comment by mjsabby** [comment]
@KodrAus I think we'd want to keep it in the LLVMSharp repo. I was referring to your scripts if they are simple enough to integrate.

We need a cmake project that'll compile the entry points.

**Comment by KodrAus** [comment]
@mjsabby Ah gotcha, if you look at the repo they are mostly simple so we should be able to just drop its contents in here somewhere and have it work. There is some organisational complexity in there though that seems like it's worth keeping separate from LLVMSharp itself and could be beneficial for other projects that want raw bindings to LLVM.

I'm happy to move in either direction though and could submit a PR that drops them in.

**Comment by KodrAus** [comment]
> We need a cmake project that'll compile the entry points.

Is this something separate from LLVMs own build project?

**Comment by tannergooding** [comment]
This issue is out of date and should be resolved with the latest binaries we produce.
